### PR TITLE
Optimize "Exposure Checks" when the app is in the foreground.

### PR DESCRIPTION
**Pull request by jeberhardt**
# Summary | Résumé

The 'ExposureCheck' is happening too frequently when the app is in the foreground.

Currently the app checks
-- before onboarding is completed
-- on the same day as the onboarding
-- every time the app is brought to the foreground

This fix ensures that onboarding has been completed for at least one day, and that it has passed the minimum time before performing another check.

# Test instructions | Instructions pour tester la modification

### CODE REVIEW

File: src/components/LastCheckedDisplay.tsx [hunk 1]
```diff
@@ -11,7 +11,7 @@ export interface LastCheckedDisplayProps {
 
 export const LastCheckedDisplay = ({textDark}: LastCheckedDisplayProps) => {
   const i18n = useI18n();
-  const [exposureStatus] = useExposureStatus();
+  const exposureStatus = useExposureStatus();
```
Comment (timarney) [inline src/components/LastCheckedDisplay.tsx:14]: This file is no longer in use.

File: src/services/ExposureNotificationService/ExposureNotificationServiceProvider.tsx [hunk 2]
```diff
@@ -55,6 +58,53 @@ export const ExposureNotificationServiceProvider = ({
     });
   }, [backgroundScheduler, exposureNotificationService]);
 
+  const status = exposureNotificationService.exposureStatus.get();
+  const storageService = useStorage();
+  const systemStatus = exposureNotificationService.systemStatus;
+
+  useEffect(() => {
+    const updateExposureStatus = async () => {
```
Comment (timarney) [inline src/services/ExposureNotificationService/ExposureNotificationServiceProvider.tsx:66]: Wondering if we can break this out into a utility function to make it easier to unit test outside of the component.

**Comment by jeberhardt** [comment]
[Optimize "Exposure Checks" when the app is in the foreground.](https://trello.com/c/7BwK8q6k/640-optimize-exposure-checks-when-the-app-is-in-the-foreground)

File: src/components/LastCheckedDisplay.tsx (hunk 1 above)
Comment (jeberhardt) [reply src/components/LastCheckedDisplay.tsx:14]: I'll remove it in a separate PR.

**Comment by jeberhardt** [comment]
[Optimize "Exposure Checks" when the app is in the foreground.](https://trello.com/c/7BwK8q6k/640-optimize-exposure-checks-when-the-app-is-in-the-foreground)

**Comment by jeberhardt** [comment]
[Optimize "Exposure Checks" when the app is in the foreground.](https://trello.com/c/7BwK8q6k/640-optimize-exposure-checks-when-the-app-is-in-the-foreground)

File: src/services/ExposureNotificationService/ExposureNotificationServiceProvider.tsx [hunk 3]
```diff
@@ -49,11 +52,43 @@ export const ExposureNotificationServiceProvider = ({
     [backendInterface, exposureNotification, i18n, secureStorage, storage],
   );
 
+  const status = exposureNotificationService.exposureStatus.get();
+  const storageService = useStorage();
+
   useEffect(() => {
     backgroundScheduler.registerPeriodicTask(async () => {
-      await exposureNotificationService.updateExposureStatusInBackground();
+      const today = new Date();
```
Comment (timarney) [inline src/services/ExposureNotificationService/ExposureNotificationServiceProvider.tsx:60]: These should use the wrapper `getCurrentDate()`

File: src/services/ExposureNotificationService/ExposureNotificationService.spec.ts [hunk 4]
```diff
@@ -428,32 +428,6 @@ describe('ExposureNotificationService', () => {
         cycleStartsAt: new OriginalDate('2020-05-14T04:10:00+0000').getTime(),
       });
     });
-
-    it('for positive', async () => {
```
Comment (timarney) [inline src/services/ExposureNotificationService/ExposureNotificationService.spec.ts:432]: Why are these removed?

File: src/services/ExposureNotificationService/ExposureNotificationServiceProvider.tsx [hunk 5]
```diff
@@ -150,3 +188,32 @@ export function useExposureNotificationSystemStatusAutomaticUpdater() {
     };
   }, [exposureNotificationService]);
 }
+
+const shouldPerformExposureCheck = (
```
Comment (timarney) [inline src/services/ExposureNotificationService/ExposureNotificationServiceProvider.tsx:192]: Let's add some unit tests for this

File: src/services/ExposureNotificationService/ExposureNotificationServiceProvider.tsx (hunk 2 above)
Comment (jeberhardt) [reply src/services/ExposureNotificationService/ExposureNotificationServiceProvider.tsx]: Moving this code from `ExposureNotificationServiceProvider.tsx` to `ExposureNotificationService.ts`.

File: src/components/LastCheckedDisplay.tsx (hunk 1 above)
Comment (timarney) [reply src/components/LastCheckedDisplay.tsx:14]: Added a PR here - https://github.com/cds-snc/covid-alert-app/pull/1218

**Comment by timarney** [comment]
@jeberhardt for the background fetch mock did you try something like 

https://github.com/transistorsoft/react-native-background-fetch/issues/204

?

**Comment by timarney** [comment]
You might want to back out of your updates and run jest --watch as you start to re-introduce your updates.

Might be easier to track down what broke and where.

Now that you know where you're headed can you break it up into multiple PRs? 

Updating the tests as you introduce the updates.  This  will also make this much easier to review.

File: src/services/ExposureNotificationService/ExposureNotificationService.ts [hunk 6]
```diff
@@ -145,14 +146,16 @@ export class ExposureNotificationService {
...
   async updateExposureStatusInBackground() {
-    await this.loadExposureStatus();
-    try {
-      captureMessage('updateExposureStatusInBackground', {exposureStatus: this.exposureStatus.get()});
-      await this.updateExposureStatus();
-      await this.processNotification();
-      captureMessage('updatedExposureStatusInBackground', {exposureStatus: this.exposureStatus.get()});
-    } catch (error) {
-      captureException('updateExposureStatusInBackground', error);
+    if (this.shouldPerformExposureCheck()) {
```
Comment (timarney) [inline src/services/ExposureNotificationService/ExposureNotificationService.ts:149]: Would prefer to do an "early return" here and avoid the "long" if statement

if(!this.shouldPerformExposureCheck()) return

File: src/services/ExposureNotificationService/ExposureNotificationService.ts [hunk 7]
```diff
@@ -228,13 +231,15 @@ export class ExposureNotificationService {
...
 
   async updateExposureStatus(): Promise<void> {
-    if (this.exposureStatusUpdatePromise) return this.exposureStatusUpdatePromise;
-    const cleanUpPromise = <T>(input: T): T => {
-      this.exposureStatusUpdatePromise = null;
-      return input;
-    };
-    this.exposureStatusUpdatePromise = this.performExposureStatusUpdate().then(cleanUpPromise, cleanUpPromise);
-    return this.exposureStatusUpdatePromise;
+    if (await this.shouldPerformExposureCheck()) {
```
Comment (timarney) [inline src/services/ExposureNotificationService/ExposureNotificationService.ts:234]: Prefer to do an early return here too

**Review by dsamojlenko — APPROVED** [review]

**Review by smcmurtry — APPROVED** [review]
